File: src/main/java/ltdjms/discord/markdown/services/DiscordMarkdownStreamProcessor.java
package ltdjms.discord.markdown.services;

import java.util.ArrayList;
import java.util.List;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import ltdjms.discord.markdown.autofix.MarkdownAutoFixer;
import ltdjms.discord.markdown.validation.MarkdownValidator;

/** 串流 Markdown 處理器：標題分段、修復、驗證與分頁。 */
public final class DiscordMarkdownStreamProcessor {

  private static final Logger LOG = LoggerFactory.getLogger(DiscordMarkdownStreamProcessor.class);

  private final MarkdownHeadingSegmenter segmenter;
  private final MarkdownValidator validator;
  private final MarkdownAutoFixer autoFixer;
  private final DiscordMarkdownSanitizer sanitizer;
  private final DiscordMarkdownPaginator paginator;

  public DiscordMarkdownStreamProcessor(
      MarkdownHeadingSegmenter segmenter,
      MarkdownValidator validator,
      MarkdownAutoFixer autoFixer,
      DiscordMarkdownSanitizer sanitizer,
      DiscordMarkdownPaginator paginator) {
    this.segmenter = segmenter;
    this.validator = validator;
    this.autoFixer = autoFixer;
    this.sanitizer = sanitizer;
    this.paginator = paginator;
  }

  public List<String> onChunk(String chunk) {
    List<String> segments = segmenter.processChunk(chunk);
    return processSegments(segments);
  }

  public List<String> flush() {
    String remaining = segmenter.drain();
    if (remaining == null || remaining.isBlank()) {
      return List.of();
    }
    return processSegments(List.of(remaining));
  }

  private List<String> processSegments(List<String> segments) {
    List<String> pages = new ArrayList<>();
    for (String segment : segments) {
      if (segment == null || segment.isBlank()) {
        continue;
      }
      String sanitizedOriginal = sanitizer.sanitize(segment);
      MarkdownValidator.ValidationResult originalValidation = validator.validate(sanitizedOriginal);
      if (originalValidation instanceof MarkdownValidator.ValidationResult.Valid) {
        pages.addAll(paginator.paginate(sanitizedOriginal));
        continue;
      }

      String fixed = autoFixer.autoFix(sanitizedOriginal);
      String sanitizedFixed = sanitizer.sanitize(fixed);
      MarkdownValidator.ValidationResult fixedValidation = validator.validate(sanitizedFixed);
      if (fixedValidation instanceof MarkdownValidator.ValidationResult.Invalid invalid) {
        LOG.warn("Markdown 分段驗證失敗: {} 個錯誤", invalid.errors().size());
      }
      pages.addAll(paginator.paginate(sanitizedFixed));
    }
    return pages;
  }
}

File: packages/ai/src/markdown/services/DiscordMarkdownStreamProcessor.ts
import type { MarkdownAutoFixer } from '../autofix/MarkdownAutoFixer.js';
import type { MarkdownValidator } from '../validation/MarkdownValidator.js';
import { isValid } from '../types.js';
import { MarkdownHeadingSegmenter } from './MarkdownHeadingSegmenter.js';
import { DiscordMarkdownSanitizer } from './DiscordMarkdownSanitizer.js';
import { DiscordMarkdownPaginator } from './DiscordMarkdownPaginator.js';

/**
 * Incremental stream markdown processor: segment, sanitize, validate, fix, paginate.
 * Matches Java DiscordMarkdownStreamProcessor.
 */
export class DiscordMarkdownStreamProcessor {
  constructor(
    private readonly segmenter: MarkdownHeadingSegmenter,
    private readonly validator: MarkdownValidator,
    private readonly autoFixer: MarkdownAutoFixer,
    private readonly sanitizer: DiscordMarkdownSanitizer,
    private readonly paginator: DiscordMarkdownPaginator,
  ) {}

  onChunk(chunk: string): string[] {
    const segments = this.segmenter.processChunk(chunk);
    return this.processSegments(segments);
  }

  flush(): string[] {
    const remaining = this.segmenter.drain();
    if (!remaining || remaining.trim().length === 0) {
      return [];
    }
    return this.processSegments([remaining]);
  }

  private processSegments(segments: string[]): string[] {
    const pages: string[] = [];
    for (const segment of segments) {
      if (!segment || segment.trim().length === 0) {
        continue;
      }

      const sanitizedOriginal = this.sanitizer.sanitize(segment);
      const originalValidation = this.validator.validate(sanitizedOriginal);
      if (isValid(originalValidation)) {
        pages.push(...this.paginator.paginate(sanitizedOriginal));
        continue;
      }

      const fixed = this.autoFixer.autoFix(sanitizedOriginal);
      const sanitizedFixed = this.sanitizer.sanitize(fixed);
      pages.push(...this.paginator.paginate(sanitizedFixed));
    }
    return pages;
  }
}
